xdpderp/xdp.c

```diff
@@ -0,0 +1,513 @@
+int xdp_prog_func(struct xdp_md *ctx) {
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data     = (void *)(long)ctx->data;
+	int is_ipv6 = -1; // -1 = unknown, 0 = no, 1 = yes
+
+	struct ethhdr *eth = data;
+	if ((void *)(eth + 1) > data_end) {
+		return end(ctx, XDP_PASS, is_ipv6);
+	}
+
+	struct iphdr *ip;
+	struct ipv6hdr *ip6;
+	struct udphdr *udp;
+
+	if (eth->h_proto == bpf_htons(ETH_P_IP)) {
+		is_ipv6 = 0;
+		ip = (void *)(eth + 1);
+		if ((void *)(ip + 1) > data_end) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		if (ip->ihl != 5 || ip->version != 4 || ip->protocol != IPPROTO_UDP) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		// validate ipv4 header checksum
+		__u32 cs_unfolded = bpf_csum_diff(0, 0, (void *)ip, sizeof(*ip), 0);
+		__u16 cs = csum_fold_flip(cs_unfolded);
+		if (cs != 0) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		udp = (void *)(ip + 1);
+		if ((void *)(udp + 1) > data_end) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		if (udp->check != 0) {
+			// https://datatracker.ietf.org/doc/html/rfc768#page-3
+			// If the computed  checksum  is zero,  it is transmitted  as all
+			// ones (the equivalent  in one's complement  arithmetic).   An all
+			// zero  transmitted checksum  value means that the transmitter
+			// generated  no checksum  (for debugging or for higher level
+			// protocols that don't care).
+			__u32 pseudo = pseudo_sum_ipv4(ip, udp->len);
+			cs = csum_const_size(pseudo, (void *)udp, data_end, MAX_UDP_LEN_IPV4);
+			if (cs != 0) {
+				return end(ctx, XDP_PASS, is_ipv6);
+			}
+		}
+	} else if (eth->h_proto == bpf_htons(ETH_P_IPV6)) {
+		is_ipv6 = 1;
+		ip6 = (void *)(eth + 1);
+		if ((void *)(ip6 + 1) > data_end) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		if (ip6->version != 6 || ip6->nexthdr != IPPROTO_UDP) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		udp = (void *)(ip6 + 1);
+		if ((void *)(udp + 1) > data_end) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+
+		// https://datatracker.ietf.org/doc/html/rfc8200#page-28
+		// Unlike IPv4, the default behavior when UDP packets are
+		// originated by an IPv6 node is that the UDP checksum is not
+		// optional.  That is, whenever originating a UDP packet, an IPv6
+		// node must compute a UDP checksum over the packet and the
+		// pseudo-header, and, if that computation yields a result of
+		// zero, it must be changed to hex FFFF for placement in the UDP
+		// header.  IPv6 receivers must discard UDP packets containing a
+		// zero checksum and should log the error.
+		__u32 pseudo = pseudo_sum_ipv6(ip6, udp->len);
+		__u16 cs = csum_const_size(pseudo, (void *)udp, data_end, MAX_UDP_LEN_IPV6);
+		if (cs != 0) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+	} else {
+		return end(ctx, XDP_PASS, is_ipv6);
+	}
+
+	__u32 config_key = 0;
+	struct config *c = bpf_map_lookup_elem(&config_map, &config_key);
+	if (!c) {
+		return end(ctx, XDP_PASS, is_ipv6);
+	}
+
+	if (bpf_ntohs(udp->dest) != c->dst_port) {
```

Comment:
This is a good idea, and will be a pain point later if we add more fields and aren't consistent on endianness. I'm not going to do it now because cilium/ebpf map encoding uses native endian, even with __be16 on the C side, and I'd rather have a native endian field on the C side + ntohs, which is more obvious/explicit than a swap on the Go side. Added a TODO in https://github.com/tailscale/tailscale/pull/12390/commits/8daf4286268aa4c7c1b0506e80f19b7b75344cfe